fendermint/app/src/app.rs

```diff
@@ -490,6 +495,45 @@ where
         Ok(response)
     }
 
+    /// Amend which transactions to put into the next block proposal.
+    async fn prepare_proposal(
+        &self,
+        request: request::PrepareProposal,
+    ) -> AbciResult<response::PrepareProposal> {
+        let txs = request.txs.into_iter().map(|tx| tx.to_vec()).collect();
+
+        let txs = self
+            .interpreter
+            .prepare((), txs)
+            .await
+            .context("failed to prepare proposal")?;
+
+        let txs = txs.into_iter().map(bytes::Bytes::from).collect();
+        let txs = take_until_max_size(txs, request.max_tx_bytes.try_into().unwrap());
```

Comment:
Same here for the gas limit check or should we only check as part of `process_proposal`?

Reply:
You are right that it must be part of both _prepare_ and _process_ otherwise you risk freaking out CometBFT by rejecting your own proposal, which is a big no-no. 

Again it would not be here as this is purely based on sizes and cannot access gas. This logic is here because it is mandatory, not something application specific like the gas, which must be delegated to interpreters.